### DISCUSSION TITLE
[java] Line length

### DISCUSSION THREAD

**Discussion by dykov** [pmd/pmd#3597] [Q&A] [answered]
Hello!
Is there any API for getting line length in Java source code (including comments).
I want to implement some rule like [checkstyle.LineLength](https://checkstyle.sourceforge.io/config_sizes.html#LineLength)
Thank you

**Comment by jsotuyod** [comment]
@dykov all `Node`s have a `getEndColumn()`.

The tricky part is mapping the node to the offending line, as nodes can be multi-line blocks. Maybe visiting all nodes looking for one that is single line && exceeds the max end column. It could still end in some duplicates to be filtered before adding violations.

I don't recall how tabs are counted towards the column count 'though, I think it's currently hardcoded to count as 4…

Also, consider that comments are not currently part of the AST. You would have to review those separately at the `CompilationUnit` level https://github.com/pmd/pmd/blob/f18ceebd7bac962ee431a2e7ab3a3458f06b624c/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTCompilationUnit.java#L32

**Comment by adangel** [comment, marked as answer]
For PLSQL, we have the rule [LineLength](https://pmd.github.io/latest/pmd_rules_plsql_codestyle.html#linelength). There we use the unparsed source code directly. In PMD6, there is no official API yet, so it would need to be integrated for Java in a similar way like in PLSQL:

Here we pass in the complete sourcecode into the root node (which is called "Input" in plsql):
https://github.com/pmd/pmd/blob/f18ceebd7bac962ee431a2e7ab3a3458f06b624c/pmd-plsql/src/main/java/net/sourceforge/pmd/lang/plsql/PLSQLParser.java#L61

The grammar needs to take it:
https://github.com/pmd/pmd/blob/f18ceebd7bac962ee431a2e7ab3a3458f06b624c/pmd-plsql/etc/grammar/PldocAST.jjt#L271
and set it into the node:
https://github.com/pmd/pmd/blob/f18ceebd7bac962ee431a2e7ab3a3458f06b624c/pmd-plsql/etc/grammar/PldocAST.jjt#L301

Then we can access the sourcecode via "node.getSourcecode()":
https://github.com/pmd/pmd/blob/f18ceebd7bac962ee431a2e7ab3a3458f06b624c/pmd-plsql/src/main/java/net/sourceforge/pmd/lang/plsql/rule/codestyle/LineLengthRule.java#L40

In PMD7, there is an official API: Each (javacc) node is also a [TextAvailableNode](https://github.com/pmd/pmd/blob/pmd/7.0.x/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/TextAvailableNode.java), so that the sourcecode is available via "node.getText()".
